src/java/org/apache/cassandra/service/AsyncProfilerService.java

```diff
@@ -0,0 +1,373 @@
+            return true;
+        }
+        catch (IllegalStateException | IllegalArgumentException e)
+        {
+            throw e;
+        }
+        catch (Throwable e)
+        {
+            logger.error("Failed to stop Async-Profiler", e);
+            return false;
+        }
+    }
+
+    public String execute(String command)
+    {
+        try
+        {
+            String result = maybeInitialize().execute(validateCommand(command));
+            logger.debug("Executed raw command in Async-Profiler: result={}, cmd={}", result, command);
+            return result;
+        }
+        catch (Throwable e)
+        {
+            logger.error("Failed to execute raw Async-Profiler command {}", command, e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    public List<String> list()
+    {
+        try
+        {
+            createLogDir();
+            return Arrays.stream(new File(logDir).list()).map(File::name).sorted().collect(toList());
+        }
+        catch (Throwable t)
+        {
+            return List.of();
+        }
+    }
+
+    public byte[] fetch(String resultFile)
+    {
+        try
+        {
+            createLogDir();
+            return Files.readAllBytes(new File(logDir, resultFile).toPath());
+        }
+        catch (Throwable t)
+        {
+            return null;
+        }
+    }
+
+    public void purge()
+    {
+        createLogDir();
+        new File(logDir).deleteRecursive();
+    }
+
+    public String status()
+    {
+        try
+        {
+            return maybeInitialize().execute("status");
+        }
+        catch (Throwable t)
+        {
+            return t.getMessage();
+        }
+    }
+
+    public boolean isEnabled()
+    {
+        return profilerInstance != null;
+    }
+
+    public static String validateOutputFileName(String outputFile)
+    {
+        if (outputFile == null || outputFile.trim().isEmpty())
+            throw new IllegalArgumentException("Output file name must not be null or empty.");
+
+        if (!VALID_FILENAME_REGEX_PATTERN.matcher(outputFile).matches())
+            throw new IllegalArgumentException(format("Output file name must match pattern %s", VALID_FILENAME_REGEX_PATTERN));
+
+        return outputFile;
+    }
+
+    public static String validateCommand(String command)
+    {
+        if (command == null || command.isBlank())
+            throw new IllegalArgumentException("Command can not be null or blank string");
+
+        return command;
+    }
+
+    /**
+     * @param duration duration of profiling
+     * @return converted string representation of duration to seconds
+     */
+    public static int parseDuration(String duration)
+    {
+        return new DurationSpec.IntSecondsBound(duration).toSeconds();
+    }
+
+    /**
+     * @throws ConfigurationException in case it is not possible to configure directory for logs.
+     */
+    private void createLogDir() throws ConfigurationException
```

Comment:
probably it is better to not name it as log directory to avoid confusion.., it contains profiles/profiling data, not profiler logs..